### PR TITLE
Caroline lucy dashboard

### LINKED ISSUE #82: User Story 16: Admin login
issueSixteen:
    title: Admin logs in
    body: >
      As an Admin
      When I log in
      Then I am redirected to "/admin/dashboard"

### LINKED ISSUE #84: User Story 18: Admin order page
issueEighteen:
    title: Admin Orders
    body: >
      As an Admin
      When I visit the dashboard
      Then I can see a listing of all orders
      And I can see the total number of orders for each status **("Ordered", "Paid", "Cancelled", "Completed")**
      And I can see a link for each individual order
      And I can filter orders to display by each status type  **("Ordered", "Paid", "Cancelled", "Completed")**
      And I have links to transition between statuses
        - I can click on "cancel" on individual orders which are "paid" or "ordered"
        - I can click on "mark as paid" on orders that are "ordered"
        - I can click on "mark as completed" on orders that are "paid"

### PR DESCRIPTION
**Pull request by lucyconklin**
Tests are passing 

Closes #84 
Closes #82

### CODE REVIEW

File: spec/features/customers/orders/customer_can_view_previous_orders_details_spec.rb [hunk 1]
```diff
@@ -26,17 +26,13 @@
   scenario "the user should see the items for the order" do
     create_orders_with_items(order_3)
     update_customer_orders(customer, order_2, order_3)
-    visit orders_path
-    click_on_order_3
+    visit customer_orders_path
+    click_on "#{order_3.id}"
```
[the remark this answers is not in the document]
Comment (CPowell23) [reply spec/features/customers/orders/customer_can_view_previous_orders_details_spec.rb]: Fixed

File: spec/features/customers/orders/customer_can_view_previous_orders_details_spec.rb [hunk 2]
```diff
@@ -7,15 +7,15 @@
...
+    expect(page).to have_current_path(customer_orders_path)
   end
 
   scenario "the user should see the details for the order" do
     create_orders_with_items(order_3)
     update_customer_orders(customer, order_2, order_3)
-    visit orders_path
-    click_on_order_3
+    visit customer_orders_path
+    click_on "#{order_3.id}"
```
[the remark this answers is not in the document]
Comment (CPowell23) [reply spec/features/customers/orders/customer_can_view_previous_orders_details_spec.rb]: Fixed

File: spec/features/admins/admin_unsuccessfully_logs_in_spec.rb [hunk 3]
```diff
@@ -0,0 +1,37 @@
+require 'rails_helper'
+
+feature "Admin sad paths" do
+
```
[the remark this answers is not in the document]
Comment (CPowell23) [reply spec/features/admins/admin_unsuccessfully_logs_in_spec.rb]: Fixed

File: spec/features/admins/admin_dashboard_shows_orders_info_spec.rb [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+    expect(page).to have_content(@order_1.date)
+    expect(page).to have_content(@order_2.date)
+    expect(page).to have_content(@order_3.date)
+  end
+
+  scenario "they can see a link to each order" do
+    expect(page).to have_link("#{@order_1.id}")
+    expect(page).to have_link("#{@order_2.id}")
+    expect(page).to have_link("#{@order_3.id}")
```
[the remark this answers is not in the document]
Comment (CPowell23) [reply spec/features/admins/admin_dashboard_shows_orders_info_spec.rb]: Fixed

File: spec/features/admins/admin_dashboard_shows_orders_info_spec.rb [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    expect(page).to have_content(@order_3.total_price_in_dollars)
+
+    expect(page).to have_content(@order_1.date)
+    expect(page).to have_content(@order_2.date)
+    expect(page).to have_content(@order_3.date)
+  end
+
+  scenario "they can see a link to each order" do
+    expect(page).to have_link("#{@order_1.id}")
+    expect(page).to have_link("#{@order_2.id}")
```
[the remark this answers is not in the document]
Comment (CPowell23) [reply spec/features/admins/admin_dashboard_shows_orders_info_spec.rb]: Fixed

File: spec/features/admins/admin_dashboard_shows_orders_info_spec.rb [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    expect(page).to have_content(@order_2.total_price_in_dollars)
+    expect(page).to have_content(@order_3.total_price_in_dollars)
+
+    expect(page).to have_content(@order_1.date)
+    expect(page).to have_content(@order_2.date)
+    expect(page).to have_content(@order_3.date)
+  end
+
+  scenario "they can see a link to each order" do
+    expect(page).to have_link("#{@order_1.id}")
```
[the remark this answers is not in the document]
Comment (CPowell23) [reply spec/features/admins/admin_dashboard_shows_orders_info_spec.rb]: Fixed

File: app/models/status.rb [hunk 7]
```diff
@@ -3,4 +3,10 @@ class Status < ApplicationRecord
   validates :name, presence: true
 
   has_many :orders
+
+  scope :sort_by_name, -> { order 'name'}
```
[the remark this answers is not in the document]
Comment (CPowell23) [reply app/models/status.rb]: Fixed

File: app/controllers/application_controller.rb [hunk 8]
```diff
@@ -16,4 +19,12 @@ def current_customer
     @current_customer ||= Customer.find(session[:customer_id]) if session[:customer_id]
   end
 
+  def current_admin?
+    !!current_admin
```
[the remark this answers is not in the document]
Comment (CPowell23) [reply app/controllers/application_controller.rb:23]: Ignore

File: app/controllers/admins_controller.rb [hunk 9]
```diff
@@ -0,0 +1,39 @@
+class AdminsController < Admins::BaseController
+  include MessageHelper
+  def show
```
[the remark this answers is not in the document]
Comment (CPowell23) [reply app/controllers/admins_controller.rb:4]: Refactored

File: app/controllers/admins/sessions_controller.rb [hunk 10]
```diff
@@ -0,0 +1,23 @@
+class Admins::SessionsController < Admins::BaseController
+  include MessageHelper
+  def new
```
[the remark this answers is not in the document]
Comment (CPowell23) [reply app/controllers/admins/sessions_controller.rb:3]: Ignore

File: spec/models/status_spec.rb [hunk 11]
```diff
@@ -8,4 +8,6 @@
   context 'associations' do
     it { is_expected.to have_many(:orders) }
   end
+
+  #need to test count_of_orders method
```
[the remark this answers is not in the document]
Comment (CPowell23) [reply spec/models/status_spec.rb]: Replaced with a passing test

File: spec/models/status_spec.rb [hunk 12]
```diff
@@ -8,4 +8,12 @@
   context 'associations' do
     it { is_expected.to have_many(:orders) }
   end
+
+  it "count_of_orders method returns the # of orders with that status" do
+    customer = create(:customer)
```
[the remark this answers is not in the document]
Comment (CPowell23) [reply spec/models/status_spec.rb]: Fixed

File: spec/features/admins/admin_can_update_admin_info_spec.rb [hunk 13]
```diff
@@ -0,0 +1,63 @@
+require "rails_helper"
+
+feature "the admin can update their own data" do
+
```
[the remark this answers is not in the document]
Comment (CPowell23) [reply spec/features/admins/admin_can_update_admin_info_spec.rb]: Fixed

File: app/controllers/admins/sessions_controller.rb [hunk 14]
```diff
@@ -0,0 +1,23 @@
+class Admins::SessionsController < Admins::BaseController
```
[the remark this answers is not in the document]
Comment (CPowell23) [reply app/controllers/admins/sessions_controller.rb:1]: Ignore

File: app/controllers/admins/orders_controller.rb [hunk 15]
```diff
@@ -0,0 +1,15 @@
+class Admins::OrdersController < ApplicationController
```
[the remark this answers is not in the document]
Comment (CPowell23) [reply app/controllers/admins/orders_controller.rb:1]: Ignore

File: app/controllers/admins/base_controller.rb [hunk 16]
```diff
@@ -0,0 +1,2 @@
+class Admins::BaseController < ApplicationController
```
[the remark this answers is not in the document]
Comment (CPowell23) [reply app/controllers/admins/base_controller.rb:1]: Ignore

File: app/assets/stylesheets/layout.sass.erb [hunk 17]
```diff
@@ -64,6 +64,9 @@ p.jumbotron-text
   width: 80%
   margin-left: auto
   margin-right: auto
+
+th.dropdown
+  display: table-cell
   margin-bottom: 60px
```
Comment (CPowell23) [inline app/assets/stylesheets/layout.sass.erb]: Was the margin-bottom: 60px supposed to go with the class above th.dropdown?

File: app/controllers/admins/orders_controller.rb [hunk 18]
```diff
@@ -0,0 +1,15 @@
...
+  def show
+    @order = Order.find(params[:order_id])
+  end
+
+  def update
+    order = Order.find(params["order_id"])
+    order.status = Status.find_by(name: params["status"])
+    order.save
+    redirect_to admin_dashboard_path
+    # if Status.all.include?(params[:status])
```
Comment (CPowell23) [inline app/controllers/admins/orders_controller.rb]: Remove comment

File: app/controllers/admins_controller.rb [hunk 19]
```diff
@@ -0,0 +1,49 @@
+class AdminsController < Admins::BaseController
+  include MessageHelper
+
+  def show
+    @admin = current_admin # take this line out
```
Comment (CPowell23) [inline app/controllers/admins_controller.rb:5]: We need this right now. We can remove it once we set up the admin authorization.

File: app/controllers/customers/orders_controller.rb [hunk 20]
```diff
@@ -1,16 +1,27 @@
...
-    @orders = current_customer.orders
+  def customer_orders
+    current_customer.orders
   end
 
   def index
+    @status_filter = params[:status_filter]
+    @statuses = Status.all
+
+    if @status_filter.nil?
```
Comment (CPowell23) [inline app/controllers/customers/orders_controller.rb:12]: Refactor the same way we did the admin show

File: app/models/order.rb [hunk 21]
```diff
@@ -11,6 +11,8 @@ class Order < ApplicationRecord
   has_many :order_items
   has_many :items, through: :order_items
 
+  scope :most_recent, -> { order 'updated_at' }
```
Comment (CPowell23) [inline app/models/order.rb:14]: Do we need to sort this by descending? We also need a test for this.

File: app/views/layouts/_nav.html.erb [hunk 22]
```diff
@@ -13,12 +13,16 @@
         <div class="logged-in-message">
           Logged in as <%=link_to "#{current_customer.first_name} #{current_customer.last_name}", dashboard_path %>
         </div>
+      <% elsif current_admin? %>
+        <div class="logged-in-message">
+          Logged in as Admin: <%=link_to "#{current_admin.first_name} #{current_admin.last_name}", admin_dashboard_path %>
```
Comment (CPowell23) [inline app/views/layouts/_nav.html.erb:18]: Can we add a method for full_name and replace this everywhere?

File: app/views/admins/show.html.erb [hunk 23]
```diff
@@ -0,0 +1,24 @@
...
+  <p>Email: <%= @admin.email %></p>
+  <p><%= link_to "Edit Profile", edit_admin_path(@admin) %></p>
+</div>
+
+<h2>Order History for all Customers</h2>
+
+<h3>Order summary</h3>
+<ul>
+<% @statuses.sort_by_name.each do |status| %>
+  <li class="status_<%= status.name.downcase %>">
```
Comment (ski-climb) [inline app/views/admins/show.html.erb:14]: Do you need this downcase?   I think the statuses are not capitalized when they're saved.

**Pull request merged**